### PR TITLE
Add "curl" module

### PR DESCRIPTION
**Pull request by falzm**
Fix previous PR (https://github.com/balabit/syslog-ng/pull/951), successfully compiled against syslog-ng master.

### CODE REVIEW

**Comment by lbudai** [comment]
Hi,

we are reviewing your PR (thanks for it!!), maybe we will put a lots of review notes here, but we will help you finishing your PR.

**Comment by falzm** [comment]
Sure, looking forward to it.

**Comment by bazsi** [comment]
I am thinking about the name. Curl is pretty widespread but shouldnt we use
"http"?

That would probably be a bit more intuitive.

**Comment by falzm** [comment]
I agree with you and have no problem at all changing the name of this module to "http", but wouldn't it conflict with the "http" Java module?

**Comment by falzm** [comment]
@lbudai I've added the init/deinit methods as found in the Java native driver (see commit c2fb835), but I still get the error at shutdown... :/

**Comment by MrAnno** [comment]
Hello @falzm,

Thank you for your contribution!

>ERROR:lib/stats/stats-cluster.c:160:stats_cluster_untrack_counter: assertion failed: (self && (self->live_mask & (1 << type)) && &self->counters[type] == (*counter))

You are getting this error because you call `g_free(self->url)` in the `_thread_deinit` function, but you also have a reference to the `url` field in `_format_stats_instance`. This function is called by syslog-ng at initialization and shutdown time as well.

I think it would be better if you moved your `g_free()` calls from `_thread_deinit` to a `free_fn` 'virtual method':

```c

static void
curl_dd_free(LogPipe *s)
{
  CurlDestinationDriver *self = (CurlDestinationDriver *)s;
  g_free(self->url);
  /* ... other deallocations */

  log_threaded_dest_driver_free(s);
}

/* ... */

LogDriver *
curl_dd_new(GlobalConfig *cfg)
{
  CurlDestinationDriver *self = g_new0(CurlDestinationDriver, 1);

  /* ... */

  self->super.super.super.super.free_fn = curl_dd_free;

  /* ... */
}
```

And because we need to read the `url` field **before** starting the thread, I would also move this if statement to `curl_dd_init`:
```c
if (!self->url) {
  self->url = g_strdup(CURL_DEFAULT_URL);
}
```

**Comment by falzm** [comment]
Thank you for your hints @MrAnno, I've fixed the issue.

**Comment by falzm** [comment]
Unfortunately I don't have much time at this moment to continue improving this module, would you consider taking it from here? As I'm not a professional developer I'm sure you'll do it much better and quicker than me... ;)

**Comment by lbudai** [comment]
Thanks for your work!
I'm not sure that we will finish the work quicker than you :-)
Any volunteer? [maybe a student who want to participate in GSoC?]

**Comment by falzm** [comment]
Unless I'm mistaken there's not much left to be done according to initial comments, maybe just the configurable `User-Agent` field? If I add this would it be OK to merge it into the main tree?

**Comment by litterbear** [comment]
@lbudai  
I'm a GSoC participant. I want to be the volunteer. 
I'll try to add the configurable User-Agent feature first.
If there are any other things that I can improve, please notify me.
I'm glad to help  ;)

File: modules/curl/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+modules/curl modules/curl/ mod-curl: modules/curl/libcurl.la
+else
+modules/curl modules/curl/ mod-curl:
+endif
+
+EXTRA_DIST        +=      \
+  modules/curl/curl-grammar.ym
+
+modules_curl_libcurl_la_CPPFLAGS  =     \
```
Comment (ihrwein) [inline modules/curl/Makefile.am]: I'd add this block inside the above `if ENABLE_CURL ... else` block.

File: modules/curl/curl-parser.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  { "password",     KW_PASSWORD },
+  { "url",          KW_URL },
+  { "headers",      KW_HEADERS },
+  { "body",         KW_BODY },
+  { NULL }
+};
+
+CfgParser curl_parser =
+{
+#if ENABLE_DEBUG
```
Comment (ihrwein) [inline modules/curl/curl-parser.c]: Please use `SYSLOG_NG_ENABLE_DEBUG` instead of `ENABLE_DEBUG`.

File: modules/curl/curl-plugin.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#ifndef SYSLOG_NG_CURL_CURL_H_INCLUDED
```
Comment (ihrwein) [inline modules/curl/curl-plugin.h]: Could you rename this define to `CURL_PLUGIN_H_INCLUDED`? That way the define names would be consistent.

File: modules/curl/curl-plugin.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+typedef struct
+{
+  LogThrDestDriver super;
+  gchar *curl;
+  gchar *url;
+  gchar *user;
+  gchar *password;
+  GList *headers;
+  gchar user_agent[1024];
```
Comment (ihrwein) [inline modules/curl/curl-plugin.h]: Could you change this field to be heap allocated? Thanks!

**Comment by litterbear** [comment]
@ihrwein  I committed my modifications here #978

**Comment by falzm** [comment]
Any idea if this module has a chance of being accepted to master branch any time soon?

**Comment by falzm** [comment]
Pull request #978 merged.

**Pull request closed without merging**